src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleKeyCache.scala

```diff
@@ -92,57 +94,124 @@ class GoogleKeyCache(
+    def oldestOf(keys: List[CachedKey]): IO[Option[String]] =
+      readFromCache(keys.minBy(_.timeCreated))
+    def newestOf(keys: List[CachedKey]): IO[Option[String]] =
+      readFromCache(keys.maxBy(_.timeCreated))
 
-      maybeActiveKey = keysFromCache.sortBy(_.timeCreated.toEpochMilli).findLast(isKeyActive(_, keysFromIam))
-      activeKey <- maybeActiveKey
-        .map { key =>
-          googleStorageAlg
-            .unsafeGetBlobBody(googleServicesConfig.googleKeyCacheConfig.bucketName, GcsBlobName(key.value))
-        }
-        .getOrElse(IO.pure(None))
+    /* segment cached keys into ideal, retired, and nascent keys
+        ideal: keys between 15 minutes and 12 days old. Use these whenever possible.
+        retired: keys older than 12 days. Avoid if possible, but prefer these over nascent keys.
+        nascent: keys newer than 15 minutes. Only use if nothing else is available; these may cause errors due to Google
+          eventual consistency.
+     */
+    val now = Instant.now()
+    val retirementTime = now.minusSeconds(googleServicesConfig.googleKeyCacheConfig.activeKeyMaxAge * (24L * 60 * 60))
+    val nascentTime = now.minusSeconds(googleServicesConfig.googleKeyCacheConfig.nascentKeyMinAgeMinutes * 60L)
+
+    val (retiredKeys, unretiredKeys) = keysInCache.partition(_.isBefore(retirementTime))
+    val (idealKeys, nascentKeys) = unretiredKeys.partition(_.isBefore(nascentTime))
+
+    if (idealKeys.nonEmpty) {
+      // if any ideal keys exist, return the newest of those
+      newestOf(idealKeys)
+    } else if (nascentKeys.nonEmpty && retiredKeys.isEmpty) {
+      // if any nascent keys exist but no retired keys exist, return the oldest nascent key
+      logger.warn(
+        s"searchCachedKeys: ${pet.id.project.value}-${pet.serviceAccount.subjectId.value} is using a nascent key"
+      )
+      oldestOf(nascentKeys)
+    } else if (nascentKeys.nonEmpty && retiredKeys.nonEmpty) {
+      // if both nascent and retired keys exist, return the newest retired key
+      logger.warn(
+        s"searchCachedKeys: ${pet.id.project.value}-${pet.serviceAccount.subjectId.value} is using a retired key; nascent keys exist"
+      )
+      newestOf(retiredKeys)
+    } else if (!withinLock) {
+      // if not within lock, return None. This will signal callers to obtain a lock and re-call this function.
+      IO.pure(None)
+    } else if (nascentKeys.isEmpty && retiredKeys.nonEmpty) {
+      // if within lock, retired keys exist, and no nascent keys exist:
+      // trigger key creation and then return newest retired key
+      for {
+        _ <- cleanupAndCreateKey(pet, keysInCache)
+        retiredKey <- newestOf(retiredKeys)
+      } yield {
+        logger.warn(
+          s"searchCachedKeys: ${pet.id.project.value}-${pet.serviceAccount.subjectId.value} is using a retired key; no nascent keys exist"
+        )
+        retiredKey
+      }
```

Comment:
in this case, I'd love to fire-and-forget the new key creation and return the retired key as fast as possible to the user. However, that would mean we release the distributed lock before new key creation finishes, which could lead to multiple keys being created in rapid succession if callers request keys multiple times.

Reply:
I don't think it is really worth it. I do think we should remove the polling in workbench libs though.